### PR TITLE
Added changing of preferred layers 

### PR DESCRIPTION
**Pull request by Astagor**
Automatically decrease video quality when showing more remote streams.

### CODE REVIEW

File: app/src/RoomClient.js [hunk 1]
```diff
@@ -2064,8 +2116,20 @@ export default class RoomClient
...
 	}
 
-	async setConsumerPreferredLayers(consumerId, spatialLayer, temporalLayer)
+	async setConsumerPreferredLayers(consumer, spatialLayer, temporalLayer)
 	{
+		const consumerId = consumer.id;
+
+		if (consumer.preferredSpatialLayer === spatialLayer &&
+			consumer.preferredTemporalLayer === temporalLayer)
+	{
```
Comment (jmencak) [inline app/src/RoomClient.js:2125]: Indentation.

File: app/src/RoomClient.js [hunk 2]
```diff
@@ -2064,8 +2116,20 @@ export default class RoomClient
...
+		logger.debug(
```
Comment (jmencak) [inline app/src/RoomClient.js:2126]: Indentation.

File: app/public/config/config.example.js [hunk 3]
```diff
@@ -158,7 +158,14 @@ var config =
 	mobileLastN          : 1,
 	// Highest number of lastN the user can select manually in 
 	// userinteface
-	maxLastN             : 5,
+	maxLastN             : 20,
+	// This values (desktop, mobile) are the limit of visible remote video streams
```
Comment (jmencak) [inline app/public/config/config.example.js:162]: Would the following make things a bit clearer?
```
// qualityDecMaxN and qualityDecMobileMaxN limit the number of visible remote video streams with
// full quality for desktop and mobile users respectively. If the number ...
```

File: app/public/config/config.example.js [hunk 4]
```diff
@@ -158,7 +158,14 @@ var config =
...
+	// with full quality. If the number of visible remote video streams is higher
+	// than this value the preferred quality will be decreased, except for
+	// the active speaker, peers pinned to spotlights, videos in new window or videos
+	// in full screen mode. This allows for having up to 20 video windows.
```
Comment (jmencak) [inline app/public/config/config.example.js:166]: Perhaps adjust the comment about the 20 video windows as it actually refers to maxLastN and not any of the new variables?  E.g.: "This allows for having more (maxLastN) video windows".

File: app/src/components/MeetingViews/Democratic.js [hunk 5]
```diff
@@ -89,26 +89,40 @@ class Democratic extends React.PureComponent
...
-		let x, y, space;
+		let x, xm, y, ym, space;
 
 		for (let rows = 1; rows <= boxes; rows = rows + 1)
 		{
-			x = width / Math.ceil(n / rows);
-			y = x / aspectRatio;
+			xm = width / Math.ceil(n / rows);
+			ym = xm / aspectRatio;
+			x = xm - 2 * 6;
```
Comment (jmencak) [inline app/src/components/MeetingViews/Democratic.js:98]: Where does `2*6` come from?  Perhaps add a comment?

**Review by jmencak — CHANGES_REQUESTED** [review]
Thank you for the PR.  This will improve the scalability with a larger number users with open cameras.

However, this is what I get when `npm run build`:

```
src/RoomClient.js
  Line 2128:1:  Expected indentation of 2 tabs but found 1  indent
  Line 2129:1:  Expected indentation of 3 tabs but found 2  indent

src/components/MeetingViews/Democratic.js
  Line 124:1:  Expected indentation of 3 tabs but found 2  indent
```
Are you sure you posted the correct version of your work?

After fixing the indentation in `src/RoomClient.js` and `src/components/MeetingViews/Democratic.js`, I was able to make this code work without issues.

File: app/src/components/MeetingViews/Democratic.js [hunk 6]
```diff
@@ -89,26 +89,40 @@ class Democratic extends React.PureComponent
...
+				if (space < 0)
+				{
+					ym = height / rows;
+					xm = aspectRatio * ym;
+					y = ym - 2 * 6;
+					x = xm - 2 * 6;
+				}
+
 				break;
 		}
```
Comment (jmencak) [inline app/src/components/MeetingViews/Democratic.js:124]: Indentation.

File: app/src/RoomClient.js (hunk 1 above)
Comment (Astagor) [reply app/src/RoomClient.js]: Fixed

File: app/src/RoomClient.js (hunk 2 above)
Comment (Astagor) [reply app/src/RoomClient.js]: Fixed

File: app/src/components/MeetingViews/Democratic.js (hunk 6 above)
Comment (Astagor) [reply app/src/components/MeetingViews/Democratic.js]: Fixed

File: app/src/RoomClient.js [hunk 7]
```diff
@@ -1979,6 +1981,56 @@ export default class RoomClient
...
+			{
+				activeVideoConsumers.push(consumer);
+			}
+		}
+
+		activeVideoConsumers.forEach((consumer) =>
+		{
+			let spatialLayer = 100;
+
+			let temporalLayer = 100;
```
Comment (jmencak) [inline app/src/RoomClient.js:2012]: Why 100 for both spacial and temporal layers?

**Review by jmencak — COMMENTED** [review]
After the lint, this works for me as is now, thank you.

File: app/src/RoomClient.js (hunk 7 above)
Comment (Astagor) [reply app/src/RoomClient.js:2012]: We need to figure out a better algorithm, I have been talking with @so010 about this. For now, it is either maximum quality (100) or lowest quality (0).

File: app/public/config/config.example.js (hunk 4 above)
Comment (Astagor) [reply app/public/config/config.example.js:166]: I think, that a general reorder of the config variable is a task we must do. All the variables regarding the number of videos should be grouped, and also the settings for simulcast, possible resolutions, as when all depend on each other. But this is a more of a strategic decision.

**Pull request closed without merging**